vault/resource_pki_backend_role.go

```diff
@@ -0,0 +1,380 @@
+	}
+	if len(domains) > 0 {
+		data["allowed_domains"] = domains
+	}
+
+	data["allow_bare_domains"] = d.Get("allow_bare_domains").(bool)
+	data["allow_subdomains"] = d.Get("allow_subdomains").(bool)
+	data["allow_glob_domains"] = d.Get("allow_glob_domains").(bool)
+	data["allow_any_name"] = d.Get("allow_any_name").(bool)
+	data["enforce_hostnames"] = d.Get("enforce_hostnames").(bool)
+	data["allow_ip_sans"] = d.Get("allow_ip_sans").(bool)
+	data["client_flag"] = d.Get("client_flag").(bool)
+	data["code_signing_flag"] = d.Get("code_signing_flag").(bool)
+	data["email_protection_flag"] = d.Get("email_protection_flag").(bool)
+	data["key_type"] = d.Get("key_type").(string)
+	data["key_bits"] = d.Get("key_bits").(int)
+
+	iUsages := d.Get("key_usage").([]interface{})
+	usages := make([]string, 0, len(iUsages))
+	for _, iUsage := range iUsages {
+		usages = append(usages, iUsage.(string))
+	}
+	if len(usages) > 0 {
+		data["key_usage"] = usages
+	}
+
+	data["use_csr_common_name"] = d.Get("use_csr_common_name").(bool)
+	data["use_csr_sans"] = d.Get("use_csr_sans").(bool)
+	data["ou"] = d.Get("ou").(string)
+	data["organization"] = d.Get("organization").(string)
+	data["generate_lease"] = d.Get("generate_lease").(bool)
+	data["no_store"] = d.Get("no_store").(bool)
+
+	_, err := client.Logical().Write(path, data)
+
+	d.SetId(path)
+
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("Error writing PKI backend role %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Wrote PKI backend role %q", path)
+
+	return pkiBackendRoleRead(d, meta)
+}
+
+func pkiBackendRoleRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	backend, err := pkiBackendRoleBackendFromPath(path)
+	if err != nil {
+		return fmt.Errorf("Invalid path %q for PKI backend role: %s", path, err)
+	}
+
+	role, err := pkiBackendRoleNameFromPath(path)
+	if err != nil {
+		return fmt.Errorf("Invalid path %q for PKI backend role: %s", path, err)
+	}
+
+	log.Printf("[DEBUG] Reading PKI backend role %q", path)
+	resp, err := client.Logical().Read(path)
+	if err != nil {
+		return fmt.Errorf("Error reading PKI backend role %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Read PKI backend role %q", path)
+	if resp == nil {
+		log.Printf("[WARN] PKI backend role %q not found, removing from state", path)
+		d.SetId("")
+		return nil
+	}
+
+	iDomains := resp.Data["allowed_domains"].([]interface{})
+	domains := make([]string, 0, len(iDomains))
+	for _, iDomain := range iDomains {
+		domains = append(domains, iDomain.(string))
+	}
+
+	iUsages := resp.Data["key_usage"].([]interface{})
+	usages := make([]string, 0, len(iUsages))
+	for _, iUsage := range iUsages {
+		usages = append(usages, iUsage.(string))
+	}
+
+	d.Set("backend", backend)
+	d.Set("role", role)
+	d.Set("ttl", resp.Data["ttl"])
+	d.Set("max_ttl", resp.Data["max_ttl"])
+	d.Set("allow_localhost", resp.Data["allow_localhost"])
+	d.Set("allowed_domains", domains)
```

Comment:
Again, sorry for setting a poor example here, but `d.Set` can actually return an error. And for primitives (strings, ints, bools, etc.) we don't bother checking it, because code gets unwieldy and there's pretty much never an error there. For complex types, though (Sets, Lists, Maps) we're moving towards checking errors. So if you could check and return the error for this and `key_usage`, that would be awesome.

Reply:
Sure, done.